File: source_auditory_transcription.m
function transcription = source_auditory_transcription(events_tsv, off_on, off_on_durations)
%Function to obtain onset times and text transcriptions of words
%played in spoken word stimulus files for auditory cohort.
%Takes in the 'events.tsv' file for a single subject and a directory to
%save the output in and prints a table (sub-A2XXX_transcription.csv) containing onset time, word, and other metrics
%from ForcedAligner. 
%off_on and off_on_durations should be provided as pre-loaded tables.
%  see the associated 'source_auditory_transcription_loop.m'


%Get list of stimulus files played to all subjects

subj_audio_files = []; %holds which audio files (number ###.wav) were played for a subject, in order
audio_start_times = []; %the time on master clock when this audio was played
pattern = '\d+(?=.wav)';

%pick out the audio files actually played to a subject
events = tdfread(events_tsv); % for example, 'sub-A2055_task-auditory_events.tsv'
for j = 1:length(string(events.type))
    if strtrim(string(events.type(j,:))) == "Sound"
        filename = events.value(j,:);
        result = regexp(filename, pattern, 'match');
        if ~isempty(result)
            extracted_number = result{1};
        else
            extracted_number = 'No match found';
        end
        subj_audio_files = [subj_audio_files; str2num(extracted_number)];
        audio_start_times = [audio_start_times;(events.onset(j))];
    end
end

%find the words and start times
relevant_events = table();
duration_table = off_on_durations.Duration;
for k = 1:length(subj_audio_files)
    % Regular expression pattern
    TGpattern = strcat(num2str(subj_audio_files(k)), '(?=.TextGrid)');
    % Apply the regular expression to each element in the cell array
    matchIndices = cellfun(@(x) ~isempty(regexp(x, TGpattern, 'once')), table2cell(off_on(:,'TextGrid')));
    % Find the indices of the matches
    matchingRowIndices = find(matchIndices);
    segment = off_on(matchingRowIndices,[4,5,9,10]);
    durations = table(duration_table(matchingRowIndices)); %dot indexing not working yet.
    durations.Properties.VariableNames = {'Duration'};
    segment = horzcat(segment,durations);
    segment.AlignOnset = segment.AlignOnset + audio_start_times(k); %aligns the onset times with the master clock
    relevant_events = vertcat(relevant_events,segment);
end




% Find the position of the first underscore and the last dot
underscoreIndex = strfind(events_tsv, '_');
dotIndex = strfind(events_tsv, '.');
% Extract the prefix (up to the first underscore)
events_tsv = char(events_tsv);
subject = events_tsv(1:underscoreIndex(1)-1);
% Construct the new file name
newFileName = [subject '_transcription.csv'];
writetable(relevant_events,fullfile(newFileName))
transcription = relevant_events;

File: source_auditory_transcription.py
# %%
import pandas as pd
from pathlib import Path
import os
import numpy as np

def source_auditory_transcription(events_tsv, audio_onsets_csv='MOUS_audio_onset_offsets_with_duration.csv', output_csv=None):
    # %%
    #Load in ForcedAligner output
    audio_onsets = pd.read_csv(audio_onsets_csv)

    # %%
    #filenames
    audio_onsets['TextGrid'] = audio_onsets['TextGrid'].apply(lambda x: x.split('/home/at/workdir/example_workbook.tg/EQ_Ramp_Int2_Int1LPF')[1].split('.TextGrid')[0])
    #audio_onsets

    # %%
    events_path = events_tsv
    events = pd.read_csv(events_path, sep='\t')
    #events

    # %%
    #block onsets
    block_onsets = events[events['value'].apply(lambda x: 'Start File' in x)]
    block_onsets = block_onsets.drop(columns=['duration'])
    #block_onsets

    # %%
    #sound file numbers
    block_onsets['value'] = block_onsets['value'].apply(lambda x: x.split('Start File ')[1].split('.wav')[0])
    #block_onsets

    # %%
    transcription = pd.DataFrame()
    for i, file in enumerate(block_onsets['value']):
        block_start = block_onsets['onset'].iloc[i]
        segment = audio_onsets[audio_onsets['TextGrid'].apply(lambda x: file in x)]
        segment = segment.drop(columns=['WordCount','NumWords','PresegBegin'])
        segment['AlignOnset'] = segment['AlignOnset'] + block_start
        #concatenate all segments
        transcription = pd.concat([transcription,segment])
    transcription = transcription.reset_index(drop=True)

    # %%
    if output_csv is not None:
        transcription.to_csv(output_csv)
    else:
        subject = Path(events_path).name.split('task-auditory_events.tsv')[0]
        output_csv = Path(events_path).parent / f'{subject}_transcription.csv'
        transcription.to_csv(output_csv)
        return transcription

    # %%



